src/helpers/request.ts

```diff
@@ -0,0 +1,85 @@
+import { ParsedUrlQueryInput } from 'querystring';
+import { addQueryToUrl, appendPathnameToUrl } from 'url-transformers';
+import { compactDefined, flow } from './fp';
+import { ApiResponse, handleFetchResponse, HandleResponse } from './response';
+import { isDefined, OmitStrict } from './typescript';
+
+type BuildUrlParams = {
+  pathname: string;
+  query?: ParsedUrlQueryInput;
+};
+
+const buildUrl = ({ pathname, query = {} }: BuildUrlParams) =>
+  flow(appendPathnameToUrl(pathname), addQueryToUrl(compactDefined(query)));
+
+type FetchParams = Pick<RequestInit, 'method'>;
+/**
+ * The params generated by the library
+ */
+type BaseRequestParams = BuildUrlParams & FetchParams;
+
+/**
+ * Additional fetch options provided by the user on a per-call basis
+ */
+type AdditionalPerFetchParams = Omit<RequestInit, keyof FetchParams>;
+type CompleteRequestParams = BaseRequestParams & AdditionalPerFetchParams;
+
+/**
+ * helper used to type-check the arguments, and add default params for all requests
+ */
+export const createRequestParams = <Args extends {}>(fn: (a: Args) => BaseRequestParams) => (
+  a: Args,
+  additionalFetchOptions: AdditionalPerFetchParams = {},
+): CompleteRequestParams => ({
+  ...fn(a),
+  ...additionalFetchOptions,
+});
+
+/**
+ * Initial parameters that apply to all calls
+ */
+type InitParams = {
+  accessKey?: string;
+  apiVersion?: string;
+  apiUrl?: string;
+} & OmitStrict<RequestInit, 'method' | 'body'>;
```

Comment:
Since we want to support both client and server use, there are two options:

- if `apiUrl` is not defined (i.e. we're calling the API directly without a proxy), then `accessKey` _must_ be defined for auth.
- if `apiUrl` is defined, then I _think_ `accessKey` must _not_ be defined since the only reason for defining the URL is to point to a proxy, in which case we want to make sure the user does not provide the `accessKey` by mistake.

We can easily do that by adjusting the types to be a union:

```ts
type InitParams = {
  apiVersion?: string;
} & OmitStrict<RequestInit, 'method' | 'body'> &
  ({ accessKey: string; apiUrl?: never } | { apiUrl: string; accessKey?: never });
```

WDYT? I'm trying to think of a scenario where someone wants to be able to provide both, but I can't think of any logical one. I'm torn between enforcing the API guidelines through the types (`accessKey` must never be exposed) vs allowing for flexibility in some exotic use-case  🤔

Reply:
> I'm trying to think of a scenario where someone wants to be able to provide both, but I can't think of any logical one.

Me neither. I like your idea! Side note: we don't really need the `?: never` properties.

Reply:
Sweet, I'll make the change.

I found that we need the `?: never` for a bunch of reasons:

- without it, TS will still allow the user to provide both, making my change pointless :confused:
```ts
type InitParams = { accessKey: string } | { apiUrl: string };
const foo = (a: InitParams) => a

foo({ apiUrl: '', accessKey: ''}) //no error
```

- without it, setting default properties (which we do for `apiUrl`) and reading the 2 values becomes more laborious, since we can no longer de-structure those two properties from the function argument right away.

Reply:
Makes sense to me 👍  I'd just make sure we're clear in whatever docs. The whole client vs server thing can be confusing cause most places wouldn't care about the client making calls to the API directly, but we have to due to the access key leaking